### PR TITLE
Add a cell-cache executor for batched generation

### PR DESCRIPTION
**Pull request by kiymetakdemir**
Implements the Executor seam over the cell KV cache. Backend-neutral: it reaches the cache through CacheBuilderRegistry and BatchControl, with the backend id, cache kind and cache-key option as configuration, so any backend registering a cell cache is served. A session is one cache sequence, and a batch is one forward carrying every input's tokens end to end on a single axis, with the cache's mask keeping the sequences apart. SessionIds are monotonic and never reissued, mapped onto cache sequence ids that do recycle; the cache frees an id once a sequence's last cell goes, which the seam forbids for a SessionId.

Read build_step first; it holds the reasoning. It flattens a batch into tokens, positions and sequence ids in one pass so entry i of each names the same token, which is how the cache pairs them when placing cells and building the mask. Sequence lengths come from the cache and do not move as inputs are laid down, so a per-sequence cursor carries the batch's own
writes.

The method loads on the first execute() rather than in create(), because a delegate may bind per-thread state as it initializes, which happens during that load, and construction is the only entry point that does not run on the engine thread.

### CODE REVIEW

File: extension/llm/batching/cell_executor.cpp [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+        const CTYPE* begin = logits.const_data_ptr<CTYPE>() + row * vocab;
+        if (params.temperature <= 0.0f) {
+          drawn = static_cast<Token>(
+              std::max_element(begin, begin + vocab) - begin);
+          return;
+        }
+        // The sampler reduces in place, so copy rather than overwrite the
+        // model's output.
+        std::vector<float> scores(begin, begin + vocab);
+        Sampler sampler(
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: Don't construct a new sampler on each execute.

We should have session_id -> SessionInfo map

And the SessionInfo can have a sampler field on it that gets set during set_sampling_params or something like that. We also don't need to do all this seed mixing.

We can just continue using the sampler object.

File: extension/llm/batching/cell_executor.cpp [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+namespace batching {
+
+using ::executorch::extension::make_tensor_ptr;
+using ::executorch::runtime::Error;
+
+namespace {
+
+// Spread one seed over positions so a session's neighbouring tokens draw
+// unrelated streams.
+std::uint64_t mix(std::uint64_t seed, std::int64_t position) {
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: Do we need mix? See comment below on sampler

File: extension/llm/batching/cell_executor.cpp [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+  sampling_.insert_or_assign(
+      session, Sampling{params, seed.value_or(nondeterministic_seed())});
+}
+
+bool CellExecutor::execute(const BatchInput& batch, BatchOutput& out) {
+  out.outputs.clear();
+  out.outputs.resize(batch.inputs.size());
+
+  // Ahead of the step: a refused load must claim no cell.
+  if (!ensure_method_loaded()) {
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: Why are we loadign in execute?

File: extension/llm/batching/cell_executor.cpp [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+      build_step(*ctl_, batch, seqs_, config_.max_session_tokens);
+  if (!step) {
+    return false;
+  }
+
+  auto tokens =
+      make_tensor_ptr({1, static_cast<int>(step->tokens.size())}, step->tokens);
+  auto positions = make_tensor_ptr(
+      {static_cast<int>(step->positions.size())}, step->positions);
+  const auto result = module_->execute(config_.method, {tokens, positions});
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: What communicates logit positions to the model?

File: extension/llm/batching/cell_executor.cpp [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+  }
+  if (result->empty() || !result->at(0).isTensor()) {
+    ET_LOG(
+        Error, "CellExecutor: %s returned no logits", config_.method.c_str());
+    return false;
+  }
+  const auto logits = result->at(0).toTensor();
+
+  for (std::size_t i = 0; i < batch.inputs.size(); ++i) {
+    const int row = step->logit_indices[i];
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp:378]: How are we communicating logit indices to the model in lm_head project?

File: extension/llm/batching/cell_executor.h [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+class CellExecutor : public Executor {
+ public:
+  struct Config {
+    cache::CacheConfig cache;
+    // Sessions open at once, and the cells each may hold. A short table refuses
+    // the whole batch and takes every session in it down, so create() rejects
+    // limits the table cannot honor and open_session() holds the count --
+    // exhaustion is kept unreachable rather than handled.
+    int max_sessions = 0;
+    int max_session_tokens = 0;
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.h]: Is max_session_tokens tied to some limit in cacheconfig?

File: extension/llm/batching/cell_executor.cpp (hunk 4 above)
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: What is the shape of result?

Is it logits per token, or only final logits?

File: extension/llm/batching/cell_executor.cpp [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+    return std::nullopt;
+  }
+
+  std::optional<Token> drawn;
+  struct {
+    [[noreturn]] void fail(Error) {
+      ET_CHECK_MSG(false, "CellExecutor: unsupported logits dtype");
+    }
+  } ctx;
+  ET_SWITCH_THREE_TYPES(
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: Can we not reuse sample_from_logits?

File: extension/llm/batching/cell_executor.h [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+  // The method is left for the first execute(). A delegate may bind per-thread
+  // state as it initializes, which happens during that load, and construction
+  // is the only entry point that does not run on the engine thread. So a
+  // method that will not load fails a batch rather than this call.
+  //
+  // nullptr = limits the cell table cannot honor, or an unregistered
+  // (backend_id, cache_kind).
+  static std::unique_ptr<CellExecutor> create(
+      std::unique_ptr<Module> module,
+      Config config);
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.h]: Just have params in create directly? Do we need another Config struct?

Config has a CacheConfig inside of it, but CacheConfig is fixed by Module and the model archetecture. Can we just construct from Module? Do we have shareable helpers for that?

File: extension/llm/batching/cell_executor.h (hunk 6 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.h]: We are checking if max_sessions * max_session_tokens > cacheconfig.capacity and reject if it exceeds capacity.

File: extension/llm/batching/cell_executor.h (hunk 8 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.h]: CacheConfig has parameters the module doesn't carry:  capacity, kv_storage_dtype, initial_capacity; so create can't build it from the module alone, the application supplies those from its own flags.

But the executor just reads one field of it (capacity, for the max_sessions × max_session_tokens check) and forwards the rest to the cache builder. 

Would it make more sense for the application to build the cache and pass it in? The executor would take capacity from the controller, and what create needs would be max_sessions, max_session_tokens, backend_id, and method.

File: extension/llm/batching/cell_executor.cpp (hunk 5 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp:378]: We don't, we talked about changing the model forward but didn't do it yet, should I do it as a follow up pr?

File: extension/llm/batching/cell_executor.cpp (hunk 4 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: It is logits per token

File: extension/llm/batching/cell_executor.cpp (hunk 7 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: We can but sample_from_logits assumes single-sequence decoding. I can either add a row parameter defaulting to last, or pass a one-row tensor view so the offset branch is skipped.

File: extension/llm/batching/cell_executor.cpp (hunk 2 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: Removing it.

File: extension/llm/batching/cell_executor.cpp (hunk 1 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: Makes sense, thank you! Sampler needs vocab_size at construction and set_sampling doesn't have it. The export script can save that to metadata. I guess the other option is building the sampler lazily on the first draw.

File: extension/llm/batching/cell_executor.cpp (hunk 3 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: It was a work around for the stream issue I got. From what I understand, the caller constructs the executor and passes it to Runner by reference, so the executor exists before the Runner, and the Runner creates the engine thread. That means create() runs on the caller's thread while every execute() runs on the engine thread.
Loading in create() registers it on the caller's thread, so the first execute() fails with "There is no Stream(gpu, N) in current thread".
But again thinking about it, a start() on Executor, called once on the engine thread before any batch should solve the issue.

File: extension/llm/batching/cell_executor.cpp (hunk 1 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.cpp]: We don't need to build sampler lazily.

The exeuctor/model should know its vocab size, and can save in private field after construction that sampler construction reads from

File: extension/llm/batching/cell_executor.cpp (hunk 3 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.cpp]: I thought your other PR where you do cache building lazily is what fixed stream issue?

File: extension/llm/batching/cell_executor.cpp (hunk 5 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.cpp:378]: We can do as follow-up PR, but I was very surprised it wasn't already being done at export stage

File: extension/llm/batching/cell_executor.cpp (hunk 7 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.cpp]: Is it difficult to call sample_from_logits per sequence in batch?

Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: No, I can do that

File: extension/llm/batching/cell_executor.h (hunk 8 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.h]: I think there are two parts to cache config. There are parts that are specific to the model and not actually configurable (these are read from pte metadata), and then there are parts that are configurable.

Maybe CacheConfig is a class with methods on it and helper that builds it from module?

```
auto config = CacheConfig::create(model) or something.
config.set_capacity(max_sessions × max_session_tokens)
```

What are your thoughts?

File: extension/llm/batching/cell_executor.cpp (hunk 5 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp:378]: I guess I just focused on single sequence that time

File: extension/llm/batching/cell_executor.h (hunk 8 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.h]: Makes sense to me, especially passing max_sessions, max_session_tokens and capacity separately and then comparing them looked redundant

File: extension/llm/batching/cell_executor.cpp (hunk 3 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: They are actually two different stream issues; one is the model's own stream created on the wrong thread, the other is the KV pool using the default stream instead of the one it was handed.

File: extension/llm/batching/executor.h [hunk 9]
```diff
@@ -67,6 +67,14 @@ class Executor {
  public:
   virtual ~Executor() = default;
 
+  // Bind whatever must live on the thread that will run the batches. Called
+  // once before any other call, on that thread.
+  //
+  // false = the executor cannot run.
+  virtual bool start() {
```
Comment (metascroy) [inline extension/llm/batching/executor.h]: I'm still not understanding this threading issue you're hitting, and why you need this. What is the exact issue?

File: extension/llm/batching/cell_executor.h [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+      std::unique_ptr<Module> module,
+      int max_sessions,
+      int max_session_tokens,
+      int kv_dtype,
+      std::string backend_id,
+      int initial_capacity = -1,
+      std::string method = "forward");
+
+  // Loads the method, naming the cache to the backend. The delegate binds
+  // per-thread state as it initializes, which happens during that load.
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.h]: Code pointer for this?

File: extension/llm/batching/cell_executor.cpp [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+    if (!token) {
+      return false;
+    }
+    out.outputs[i] = Output{session, {*token}};
+  }
+  return true;
+}
+
+std::optional<Token> CellExecutor::sample_row(
+    ::executorch::aten::Tensor& logits,
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp:393]: Can we use the non-aten tensor?

File: extension/llm/batching/cell_executor.cpp [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+      vocab_size_(vocab_size) {}
+
+CellExecutor::~CellExecutor() = default;
+
+std::unique_ptr<CellExecutor> CellExecutor::create(
+    std::unique_ptr<Module> module,
+    int max_sessions,
+    int max_session_tokens,
+    int kv_dtype,
+    std::string backend_id,
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: Doesn't module know the backend?

File: extension/llm/batching/cell_executor.cpp (hunk 5 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.cpp:378]: But even for single sequence, I expected last token selection, but it seems that isn't being done

File: extension/llm/batching/cell_executor.cpp [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+  if (module->load() != Error::Ok) { // a no-op once the caller has loaded it
+    ET_LOG(Error, "CellExecutor: the program did not load");
+    return nullptr;
+  }
+
+  auto cfg = cache::et::config_from_program(*module);
+  if (!cfg.ok()) {
+    return nullptr;
+  }
+  cfg->capacity = max_sessions * max_session_tokens;
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp:201]: How does capacity differ from initial_capacity?

If capacity dynamically grows, why do we need to set it to max_sessions * max_session_tokens?

File: extension/llm/batching/cell_executor.cpp [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+      std::move(cache),
+      std::move(session),
+      max_sessions,
+      max_session_tokens,
+      std::move(backend_id),
+      std::move(method),
+      logits_sizes[logits_sizes.size() - 1]));
+}
+
+bool CellExecutor::start() {
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp]: this looks like it should be called load?

File: extension/llm/batching/cell_executor.cpp (hunk 3 above)
Comment (metascroy) [reply extension/llm/batching/cell_executor.cpp]: What is a minimal repro for the second model stream on wrong thread?

That seems like a bigger UI issue that I don't want to paper over with a short term fix.

File: extension/llm/batching/cell_executor.cpp [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+
+  std::vector<std::int32_t> seq_ids;
+  seq_ids.reserve(total);
+  // Truncations the batch asks for, held until every input has been checked.
+  std::vector<std::pair<std::int32_t, int>> rewinds;
+  // Where each sequence stands mid-batch: the cache still reports what it held
+  // before the step, so the batch's own writes live here.
+  std::unordered_map<std::int32_t, int> cursor;
+
+  for (const Input& input : batch.inputs) {
```
Comment (metascroy) [inline extension/llm/batching/cell_executor.cpp:66]: Does model have a max accepted size via prefill chunk size or something?

Are we overflowing that limit?

File: extension/llm/batching/cell_executor.h (hunk 10 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.h]: Module::load_method
  → Method::init                       runtime/executor/method.cpp:883
      → BackendDelegate::Init          runtime/executor/method.cpp:974
          → backend->init(...)         runtime/executor/method.cpp:123
              → new (handle) MLXHandle()          MLXBackend.cpp:230
                  → new_stream(Device::gpu)       MLXBackend.cpp:194

File: extension/llm/batching/cell_executor.cpp (hunk 13 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp:201]: capacity is the ceiling so it doesn't grow, initial_capacity is the starting allocation. I set it to the product because that's the most the executor can ever need.

File: extension/llm/batching/cell_executor.cpp (hunk 12 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: I now read it from the program and dropped the parameter, rejecting a method spanning more than one backend.

File: extension/llm/batching/cell_executor.cpp (hunk 14 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp]: Removed this after the mlx backend fix.

File: extension/llm/batching/cell_executor.cpp (hunk 11 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp:393]: executorch::aten::Tensor is the build-mode alias, not ATen; it's torch::executor::Tensor unless USE_ATEN_LIB is set, so a default build links no ATen. It's also what EValue::toTensor() returns and what sample_from_logits takes.

File: extension/llm/batching/cell_executor.cpp (hunk 15 above)
Comment (kiymetakdemir) [reply extension/llm/batching/cell_executor.cpp:66]: Yes, the export bounds the traced seq_len, and we were overflowing it. build_step now refuses an over-wide batch, and max_step_tokens() exposes the bound so a caller can size the scheduler's max_batch_tokens by it.
